examples_tests/61.UI/UI_System.cpp

```diff
@@ -0,0 +1,1102 @@
+
+		//if (io.ConfigFlags & ImGuiConfigFlags_NoMouseCursorChange)
+		//{
+		//    return;
+		//}
+		//ImGuiMouseCursor imgui_cursor = ImGui::GetMouseCursor();
+		//if (io.MouseDrawCursor || imgui_cursor == ImGuiMouseCursor_None)
+		//{
+		//    // Hide OS mouse cursor if imgui is drawing it or if it wants no cursor
+		//    MSDL::SDL_ShowCursor(MSDL::SDL_FALSE);
+		//}
+		//else
+		//{
+		//    // Show OS mouse cursor
+		//    MSDL::SDL_SetCursor(state->mouseCursors[imgui_cursor] ? state->mouseCursors[imgui_cursor] : state->mouseCursors[ImGuiMouseCursor_Arrow]);
+		//    MSDL::SDL_ShowCursor(MSDL::SDL_TRUE);
+		//}
+	//}
+
+	//-------------------------------------------------------------------------------------------------
+
+	bool Render(IGPUCommandBuffer& commandBuffer, int frameIndex)
+	{
+		ImGuiIO& io = ImGui::GetIO();
+		assert(io.Fonts->IsBuilt() && "Font atlas not built! It is generally built by the renderer backend. Missing call to renderer _NewFrame() function? e.g. ImGui_ImplOpenGL3_NewFrame().");
+
+		UpdateMousePositionAndButtons();
+		
+		// Setup desired Vulkan state
+		// Bind pipeline and descriptor sets:
+		commandBuffer.bindGraphicsPipeline(state->pipeline.get());
+		commandBuffer.bindDescriptorSets(EPBP_GRAPHICS, state->independentPipeline->getLayout(), 0, 1, &state->gpuDescriptorSet.get());
+		
+		auto const* drawData = ImGui::GetDrawData();
+
+		if (drawData == nullptr)
+		{
+			return false;
+		}
+		
+		// Avoid rendering when minimized, scale coordinates for retina displays (screen coordinates != framebuffer coordinates)
+		float const frameBufferWidth = drawData->DisplaySize.x * drawData->FramebufferScale.x;
+		float const frameBufferHeight = drawData->DisplaySize.y * drawData->FramebufferScale.y;
+		if (frameBufferWidth > 0 && frameBufferHeight > 0 && drawData->TotalVtxCount > 0)
+		{
+			// Create or resize the vertex/index buffers
+			size_t const vertexSize = drawData->TotalVtxCount * sizeof(ImDrawVert);
+			size_t const indexSize = drawData->TotalIdxCount * sizeof(ImDrawIdx);
+
+			IGPUBuffer::SCreationParams vertexCreationParams = {};
+			vertexCreationParams.usage = nbl::asset::IBuffer::EUF_VERTEX_BUFFER_BIT;
+			vertexCreationParams.canUpdateSubRange = true;
+
+			auto & vertexBuffer = state->vertexBuffers[frameIndex];
+
+			if (vertexBuffer.isValid() == false || vertexBuffer->getSize() < vertexSize)
+			{
+				vertexBuffer = state->device->createCPUSideGPUVisibleGPUBufferOnDedMem(
+					vertexCreationParams,
+					vertexSize
+				);
+			}
+
+			IGPUBuffer::SCreationParams indexCreationParams = {};
+			indexCreationParams.usage = nbl::asset::IBuffer::EUF_INDEX_BUFFER_BIT;
+			indexCreationParams.canUpdateSubRange = true;
+
+			auto & indexBuffer = state->indexBuffers[frameIndex];
+
+			if (indexBuffer.isValid() == false || indexBuffer->getSize() < indexSize)
+			{
+				indexBuffer = state->device->createCPUSideGPUVisibleGPUBufferOnDedMem(
+					indexCreationParams,
+					indexSize
+				);
+			}
```

Comment:
I like the fact you're round robining buffers, and not re-creating every frame.

However we have an "Upstreaming" streaming buffer in `IUtilities`

Its basically a persistently mapped host visible buffer, which you can suballocate from.
The nice part is that you can "bind" and draw directly from it.
You just need to remember to flush.